### PR TITLE
dpcpp: added a new package

### PR DESCRIPTION
**Pull request by ravil-mobile**
* dpcpp stands for Data Parallel C++
* dpcpp is Intel's implementation of SYCL
* tested with cuda. According to Intel Docs, only cuda@10.2 is supported
* ROCm has not been tested. I suggeset to focus on it in the next
contribution

### CODE REVIEW

**Comment by ravil-mobile** [comment]
I've just realized that I need to append env. variables to include `headers`. I will push an update

**Review by alalazo — CHANGES_REQUESTED** [review]
Just read your comment. Feel free to ping me when you think this is ready for a review

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+from spack import *
+
+
+class Dpcpp(CMakePackage):
+    """Data Parallel C++ compiler: Intel's implementation of SYCL programming model"""
+
+    homepage = 'https://intel.github.io/llvm-docs/'
+    version('develop', git='https://github.com/intel/llvm.git', branch='sycl', submodules=True)
+    version('stable', git='https://github.com/intel/llvm.git', commit='7981e6027a8cb43028be55d19f12cb1e30a28f5a', submodules=True)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dpcpp/package.py]:
```suggestion
    git = 'https://github.com/intel/llvm.git'
    
    version('develop', branch='sycl', submodules=True)
    version('stable', commit='7981e6027a8cb43028be55d19f12cb1e30a28f5a', submodules=True)
```

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    variant('arm', default=False, description='build ARM support rather than x86')
+    variant('disable-esimd-cpu', default=False, description='build without ESIMD_CPU support')
+    variant('no-assertions', default=False, description='build without assertions')
+    variant('docs', default=False, description='build Doxygen documentation')
+    variant('no-werror', default=False, description='Do not treat warnings as errors')
+    variant('shared-libs', default=False, description='Build shared libraries')
+    variant('use-libcxx', default=False, description='build sycl runtime with libcxx')
+    variant('use-lld', default=False, description='Use LLD linker for build')
+
+    depends_on('cmake@3.16.2:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dpcpp/package.py]:
```suggestion
    depends_on('cmake@3.16.2:', type='build')
```

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    depends_on('ninja@1.10.0:', type='build')
+
+    depends_on('cuda@10.2.89', when='+cuda', type=('build', 'link', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dpcpp/package.py]:
```suggestion
    depends_on('cuda@10.2.0:10.2.999', when='+cuda', type=('build', 'link', 'run'))
```

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 1 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: done

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 2 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: done

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 3 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: done

**Comment by ravil-mobile** [comment]
Hi again, 

I used the latest changes in the script, installed `dpcpp` with Spack and tested it using this [example](https://github.com/oneapi-src/oneAPI-samples/blob/master/DirectProgramming/DPC%2B%2B/DenseLinearAlgebra/vector-add/src/vector-add-usm.cpp) with `cuda` back-end. Everything is ok with respect to this particular example. However, there are some problems with linking i.e., when I tried to compile something more complex. Sometimes setting `CXXFLAGS=-stdlib=libstdc++` helps. But it is a general problem which occurs while linking `clang` and `gcc` libs together.

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+        return args
+
+    def setup_run_environment(self, env):
+        bin_path = self.spec.prefix.bin
+        for env_var_name, compiler in zip(['CC', 'CXX'], ['clang', 'clang++']):
+            env.set(env_var_name, os.path.join(bin_path, compiler))
+
+        env.set('FC', str(self.compiler.fc) if self.compiler.fc else '')
+        env.set('F77', str(self.compiler.f77) if self.compiler.f77 else '')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dpcpp/package.py]: I think you want to remove these. These will affect all software builds that anyone does after loading this module. It's better to let users pick their own compiler.

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+        include_env_vars = ['C_INCLUDE_PATH', 'CPLUS_INCLUDE_PATH', 'INCLUDE']
+        for var in include_env_vars:
+            env.prepend_path(var, self.prefix.include)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dpcpp/package.py:135]: We used to set these by default for all packages. @alalazo do you remember why we got rid of this?

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 4 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Hi,

`dpcpp` is essentially clang-13 compiler. I tool this trick from `gcc` package: https://github.com/spack/spack/blob/8867827a89bffea655031c839a2f88c5fc6ae367/var/spack/repos/builtin/packages/gcc/package.py#L670-L683

However, I agree with you  with respect to `fortran` compiler. I think lines 115 and 116 can be removed. What do you think?

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: There is `flang` project in llvm-intel tree ([here](https://github.com/intel/llvm)). I don't know maybe it makes to build it by default and set `FC` and `F77` to that one.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: That sounds good to me, didn't realize this package is basically a compiler.

**Comment by ravil-mobile** [comment]
I've just figure out that it would be nice to add the following to `setup_run_environment`: 
```
        sycl_build_pi_rocm_platform = self.spec.variants['rocm-platform'].value
        if '+cuda' in self.spec or sycl_build_pi_rocm_platform == 'cuda':
            env.prepend_path(var, self.spec['cuda'].prefix.bin)
```

Otherwise, users always need to  load a specific `cuda` version by themselves

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        for env_var_name, compiler in zip(['CC', 'CXX'], ['clang', 'clang++']):
+            env.set(env_var_name, os.path.join(bin_path, compiler))
+
+        include_env_vars = ['C_INCLUDE_PATH', 'CPLUS_INCLUDE_PATH', 'INCLUDE']
+        for var in include_env_vars:
+            env.prepend_path(var, self.prefix.include)
+
+        sycl_build_pi_rocm_platform = self.spec.variants['rocm-platform'].value
+        if '+cuda' in self.spec or sycl_build_pi_rocm_platform == 'cuda':
+            env.prepend_path(var, self.spec['cuda'].prefix.bin)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dpcpp/package.py]: What is `var` here? Should this be `PATH`?

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Yes, you're correct. Sorry, it was unsuccessful copy&paste. I fixed it

**Comment by ravil-mobile** [comment]
Hi @alalazo and @adamjstewart, 

I've just re-based this branch with `develop`. Are there some issues to discuss?

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    variant('cuda', default=False, description='switch from OpenCL to CUDA')
+    variant('rocm', default=False, description='switch from OpenCL to ROCm')
+    variant('rocm-platform', default='AMD', values=('AMD', 'NVIDIA'), multi=False, description='choose ROCm backend')
+    variant('arm', default=False, description='build ARM support rather than x86')
+    variant('disable-esimd-cpu', default=False, description='build without ESIMD_CPU support')
+    variant('no-assertions', default=False, description='build without assertions')
+    variant('docs', default=False, description='build Doxygen documentation')
+    variant('no-werror', default=False, description='Do not treat warnings as errors')
+    variant('shared-libs', default=False, description='Build shared libraries')
+    variant('use-lld', default=False, description='Use LLD linker for build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dpcpp/package.py]:
```suggestion
    variant('esimd_cpu', default=True, description='build with ESIMD_CPU support')
    variant('assertions', default=True, description='build with assertions')
    variant('docs', default=False, description='build Doxygen documentation')
    variant('shared', default=False, description='Build shared libraries')
    variant('lld', default=False, description='Use LLD linker for build')
```
I prefer to have variants with positive names. Otherwise you end up with double negatives like `~disable-...`, which makes a lot less sense than `+enable-...`. Also, things like `no-werror` do not affect what gets installed, only whether or not it can be built. For flags like these, it's better to add them unconditionally, or conditionally on the compiler. `+shared` is what we use in other packages. Let me know if you have any questions.

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Agree. I took these options directly from `intel/llvm` 

https://github.com/intel/llvm/blob/849774e32d01f9af8778524bed5656a2d72065d7/buildbot/configure.py#L159-L189

I will rework.

Btw, @adamjstewart, what do you think about the package name? Should it stay as `dpcpp` or it is better to rename to something `intel-dpcpp`?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: No preference on package name. Ask yourself if there are any other dpcpp libraries other than Intel's. Also ask yourself what people usually refer to the software as.

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: > Ask yourself if there are any other dpcpp libraries other than Intel's

You are right. The official name is dpcpp.

> Also ask yourself what people usually refer to the software as.

Some people may think that it should be `oneapi`. But `oneapi` is a suite of Intel tools. I also asked some Intel people and they recommended me to stick to `dpcpp` name

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Hi @adamjstewart 
I also renamed variants as you suggested. Moreover, I added  
```
        if self.compiler.name == 'gcc':
            gcc_prefix = ancestor(self.compiler.cc, 2)
            args.append(self.define('GCC_INSTALL_PREFIX', gcc_prefix))
```

This is supposed to resolve linking problems caused by `libstd++`. I tested it on my system with one of my projects

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    variant('cuda', default=False, description='switch from OpenCL to CUDA')
+    variant('rocm', default=False, description='switch from OpenCL to ROCm')
+    variant('rocm-platform', default='AMD', values=('AMD', 'NVIDIA'), multi=False, description='choose ROCm backend')
+    variant('arm', default=False, description='build ARM support rather than x86')
+    variant('esimd-cpu', default=False, description='build with ESIMD_CPU support')
+    variant('assertions', default=False, description='build with assertions')
+    variant('docs', default=False, description='build Doxygen documentation')
+    variant('werror', default=False, description='Treat warnings as errors')
+    variant('shared', default=False, description='Build shared libraries')
+    variant('use-lld', default=False, description='Use LLD linker for build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dpcpp/package.py]:
```suggestion
    variant('lld', default=False, description='Use LLD linker for build')
```

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: A good point. Done.

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            self.define('LLVM_EXTERNAL_SYCL_SOURCE_DIR', sycl_dir),
+            self.define('LLVM_EXTERNAL_LLVM_SPIRV_SOURCE_DIR', spirv_dir),
+            self.define('LLVM_EXTERNAL_XPTI_SOURCE_DIR', xpti_dir),
+            self.define('XPTI_SOURCE_DIR', xpti_dir),
+            self.define('LLVM_EXTERNAL_XPTIFW_SOURCE_DIR', xptifw_dir),
+            self.define('LLVM_EXTERNAL_LIBDEVICE_SOURCE_DIR', libdevice_dir),
+            self.define('LLVM_ENABLE_PROJECTS', llvm_enable_projects),
+            self.define('LIBCLC_TARGETS_TO_BUILD', libclc_targets_to_build),
+            self.define('SYCL_BUILD_PI_CUDA', 'ON' if is_cuda else 'OFF'),
+            self.define('SYCL_BUILD_PI_ROCM', 'ON' if is_rocm else 'OFF'),
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dpcpp/package.py]:
```suggestion
            self.define_from_variant('SYCL_BUILD_PI_CUDA', 'cuda'),
            self.define_from_variant('SYCL_BUILD_PI_ROCM', 'cuda'),
```

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Done,

Changed to 
```
            self.define_from_variant('SYCL_BUILD_PI_CUDA', 'cuda'),
            self.define_from_variant('SYCL_BUILD_PI_ROCM', 'rocm'),
```

**Review by adamjstewart — DISMISSED** [review]
This looks good to me now. Still hoping for a response from @alalazo about the environment variables.

**Comment by ravil-mobile** [comment]
Thanks @adamjstewart. For some reasons, I don't have a button beside @alalazo on the top right to re-request review

**Comment by adamjstewart** [comment]
Same, weird

**Comment by ravil-mobile** [comment]
But in any case I will need to set `include` env. variables to `env.prepend_path(var, self.prefix.include.sycl)`

All oneapi [sample codes](https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/DPC%2B%2B) starts with 
```
#include <CL/sycl.hpp>
```
which is in `<dpcpp>/include/sycl`

the structure of the include dir. in dpcpp installation folder is the following
```
$ tree -L 2 $(spack location -i dpcpp)/include
├── level_zero
│   ├── layers
│   ├── loader
│   ├── ze_api.h
│   ├── ze_ddi.h
│   ├── zes_api.h
│   ├── zes_ddi.h
│   ├── zet_api.h
│   └── zet_ddi.h
└── sycl
    ├── CL
    ├── ext
    └── sycl.hpp
```

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    # NOTE: AMD HIP needs to be tested; it will be done in the next update
+    # depends_on('cuda@10.2.0:10.2.999', when='rocm-platform=NVIDIA', type='build')
+    # depends_on('hip@4.0.0:', when='+rocm', type='build')
+
+    build_targets = ['deploy-sycl-toolchain']
+    install_targets = ['deploy-sycl-toolchain']
+
+    root_cmakelists_dir = 'llvm'
+
+    patch('cuda-backend.patch', when='@stable +cuda')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dpcpp/package.py]: Can you add a comment on top of this patch directive to explain provenance of the patch and what it does / why it is needed? Bonus if you can get rid of the patch and fetch from a url (but not a requirement).

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    variant('arm', default=False, description='build ARM support rather than x86')
+    variant('esimd-cpu', default=False, description='build with ESIMD_CPU support')
+    variant('assertions', default=False, description='build with assertions')
+    variant('docs', default=False, description='build Doxygen documentation')
+    variant('werror', default=False, description='Treat warnings as errors')
+    variant('shared', default=False, description='Build shared libraries')
+    variant('lld', default=False, description='Use LLD linker for build')
+
+    depends_on('cmake@3.16.2:', type='build')
+    depends_on('ninja@1.10.0:', type='build')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dpcpp/package.py:34]: @adamjstewart Don't we need:
```
generator = 'Ninja'
```
as an attribute at the class level?

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+    build_targets = ['deploy-sycl-toolchain']
+    install_targets = ['deploy-sycl-toolchain']
+
+    root_cmakelists_dir = 'llvm'
+
+    patch('cuda-backend.patch', when='@stable +cuda')
+
+    def cmake_args(self):
+        llvm_external_projects = 'sycl;llvm-spirv;opencl;libdevice;xpti;xptifw'
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dpcpp/package.py:48]: Does this mean that `CMake` will look for external project and try to build them outside of Spack's control? If so, is there a way to add the dependencies in Spack and avoid the external projects?

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+        sycl_dir = os.path.join(self.stage.source_path, 'sycl')
+        spirv_dir = os.path.join(self.stage.source_path, 'llvm-spirv')
+        xpti_dir = os.path.join(self.stage.source_path, 'xpti')
+        xptifw_dir = os.path.join(self.stage.source_path, 'xptifw')
+        libdevice_dir = os.path.join(self.stage.source_path, 'libdevice')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dpcpp/package.py:54]: I guess this answer one of my questions :slightly_smiling_face:

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    git = 'https://github.com/intel/llvm.git'
+
+    version('develop', branch='sycl')
+    version('stable', commit='7981e6027a8cb43028be55d19f12cb1e30a28f5a')
+
+    maintainers = ['ravil-mobile']
+    variant('cuda', default=False, description='switch from OpenCL to CUDA')
+    variant('rocm', default=False, description='switch from OpenCL to ROCm')
+    variant('rocm-platform', default='AMD', values=('AMD', 'NVIDIA'), multi=False, description='choose ROCm backend')
+    variant('arm', default=False, description='build ARM support rather than x86')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dpcpp/package.py]: I think a check on the target would be more consistent that a variant.

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    variant('esimd-cpu', default=False, description='build with ESIMD_CPU support')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dpcpp/package.py:25]: Same here (better a check on `target` maybe)

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        xptifw_dir = os.path.join(self.stage.source_path, 'xptifw')
+        libdevice_dir = os.path.join(self.stage.source_path, 'libdevice')
+        llvm_targets_to_build = 'X86'
+        llvm_enable_projects = 'clang;' + llvm_external_projects
+        libclc_targets_to_build = ''
+        sycl_build_pi_rocm_platform = self.spec.variants['rocm-platform'].value
+
+        # replace not append, so ARM ^ X86
+        if '+arm' in self.spec:
+            llvm_targets_to_build = 'ARM;AArch64'
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dpcpp/package.py]: The idea above is to avoid variants and use something like:
```suggestion
        if self.spec.satisfies('target=aarch64:'):
            llvm_targets_to_build = 'ARM;AArch64'
```

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+        is_cuda = '+cuda' in self.spec
+        is_rocm = '+rocm' in self.spec
+
+        if is_cuda or is_rocm:
+            llvm_enable_projects += ';libclc'
+
+        if is_cuda:
+            llvm_targets_to_build += ';NVPTX'
+            libclc_targets_to_build = 'nvptx64--;nvptx64--nvidiacl'
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dpcpp/package.py:75]: These seems all dependencies __if__ this package depended on `llvm`

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 5 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/dpcpp/package.py:135]: I think because it's bad practice in C type languages. If you set these environment variables for every package you load you have more opportunities for name collisions. During builds we explicitly _sanitize_ for these variables:

https://github.com/spack/spack/blob/a3d8e95e76e50cd5e9720295902617d56ab28553/lib/spack/spack/build_environment.py#L163-L168

Can we remove these variables here?

File: var/spack/repos/builtin/packages/dpcpp/package.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        if self.compiler.name == 'gcc':
+            gcc_prefix = ancestor(self.compiler.cc, 2)
+            args.append(self.define('GCC_INSTALL_PREFIX', gcc_prefix))
+
+        return args
+
+    def setup_run_environment(self, env):
+        bin_path = self.spec.prefix.bin
+        for env_var_name, compiler in zip(['CC', 'CXX'], ['clang', 'clang++']):
+            env.set(env_var_name, os.path.join(bin_path, compiler))
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dpcpp/package.py:131]: I don't think it's good to set these here. If you have an environment with more than one compiler this package would set itself as a default.

**Review by alalazo — CHANGES_REQUESTED** [review]
I left a few comments, let me know what you think.

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 11 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/dpcpp/package.py:34]: Yes

**Comment by alalazo** [comment]
Also pinging  @rscohn2 in case he has opinions / he is interested since he usually maintains Intel related packages.

**Comment by rscohn2** [comment]
@pvchupin: This is a spack pacakge for open source dpcpp. It automates download/configure/build/install and makes it easy when you are using spack to manage your other dependencies.  We have a separate package to install product releases. Can you ask someone to look this over--mainly for cmake and build options.

**Comment by pvchupin** [comment]
> @pvchupin: This is a spack pacakge for open source dpcpp. It automates download/configure/build/install and makes it easy when you are using spack to manage your other dependencies. We have a separate package to install product releases. Can you ask someone to look this over--mainly for cmake and build options.

Tagging @bader  and @vladimirlaz

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 10 above)
Comment (bader) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: +1. Considering that it's applied only for "stable" version, I assume that this problem is already resolved on the tip of the branch. If so, I suggest we take the stable version close to the tip of the branch, to avoid additional patches.

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 12 above)
Comment (bader) [reply var/spack/repos/builtin/packages/dpcpp/package.py:48]: > Does this mean that `CMake` will look for external project and try to build them outside of Spack's control? 

No. The source code for these "projects" are located in the same repository. The DPC++ project is forked from LLVM project and these new libraries are integrated into LLVM's CMake build system as "external project" i.e. their build is disabled by default. To enable build of these projects, developers should pass additional CMake variable `LLVM_EXTERNAL_PROJECTS` with the list of the projects to enable.

More on can be found in LLVM's documentation: https://llvm.org/docs/CMake.html#llvm-related-variables (look for LLVM_EXTERNAL_PROJECTS).

@ravil-mobile, is it possible to use `./buildbot/configure.py` script to generate Ninja files instead of using CMake directly? It might be difficult to maintain as our Get Started Guide recommends using this script, we are focused on keeping `configure.py` defaults stable, but there might be breaking changes in CMake options configuration.

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py:48]: Hi @bader, 

> is it possible to use ./buildbot/configure.py script to generate Ninja files instead of using CMake directly?

I don't know. It is better to ask Spack experts like @alalazo  and @adamjstewart

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 13 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py:54]: yes, as @bader mentioned it is a fork of llvm)

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 14 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Yep, agree. It is a good idea

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 18 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py:131]: Dpcpp is an expended C/C++ compiler. If a user loads this package then he/she definitely wants to use dpcpp to compile their projects with dpcpp. I took this from GCC in Spack.

https://github.com/spack/spack/blob/2d78045cdde0b03b8f61b82ab249cd75c7514d35/var/spack/repos/builtin/packages/gcc/package.py#L667-L682    

this line `env.set(lang.upper(), abspath) `

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 10 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Hi, @bader and @alalazo,

I just want to mention that I am not a developer of dpcpp. I am just a user who always wanted to have a reproducible installation of this compiler. 

To be honest, I cannot figure out versioning of dpcpp (see, [here](https://github.com/intel/llvm/releases) and [there](https://github.com/intel/llvm/tags)). There are lot's of tags (nightly builds) and it is not clear what is stable and what is not. `stable` version in Spack refers to just a nightly build (that I took a month ago) that didn't fail any test. I just needed a reference to something which is not changing e.g., not a tip of the master branch.  However, I've just noticed that there is a **really** stable release ([this one](https://github.com/intel/llvm/releases/tag/2021-07)). So I will change the reference. 

Talking about patch, honestly the solution came from [this thread](https://github.com/intel/llvm/issues/4307). I didn't go into details what caused the problem. But because I believe that the bug will be fixed in the future I didn't apply the patch to the develop branch.

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 17 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py:75]: it is, in fact, llvm

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 10 above)
Comment (Michoumichmich) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Hello, the patch I posted there doesn't work anymore with https://github.com/intel/llvm/commit/3c7971d77de6a4f7cdf5831f4fb909391afd89f3

Comment (Michoumichmich) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: but the bug is still present if you compile with `nvptx64-nvidia-cuda`

Comment (Michoumichmich) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: @ravil-mobile You can compile with `nvptx64-nvidia-cuda--sm_50` on the current branch, without using the patch

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Hi, @Michoumichmich,

Do you mean I can take `dpcpp` from the tip of the master, compile it, compile my application (or a test application) with `-fsycl -fsycl-targets=nvptx64-nvidia-cuda--sm_50`, and it will work without the patch?

What about [this release](https://github.com/intel/llvm/releases/tag/2021-07)? Do I still need a patch for this? I mean a Spack package can have a `develop` version which points to the tip of the master but there should be at least one stable version (or at least something which people can reference as a stable version).

Comment (Michoumichmich) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: All the issues I had with dpcpp that needed that patch are gone on the master branch and the target `nvptx64-nvidia-cuda--sm_whatever` and `nvptx64-nvidia-cuda` (seems to work now too). I don't know if the release version works without the patch, but it works with it. Maybe you could take the release and patch on top of it the commit mentioned ? It could work. I can't rebuild llvm right now
Here's a code that was breaking before: https://github.com/Michoumichmich/SYCL-Hashing-Algorithms  
To sum up, what you were doing works as long as you're not pulling from master but from the release you mentioned

Comment (Michoumichmich) [reply var/spack/repos/builtin/packages/dpcpp/package.py:47]: My apologies for being confusing

Comment (Michoumichmich) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: @ravil-mobile Hi, I just checked and the release you mentioned still needs the patch, without it the executable fails with CUDA_ERROR_NOT_FOUND

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: > Hi, I just checked and the release you mentioned still needs the patch, without it the executable fails with CUDA_ERROR_NOT_FOUN

Ok, thanks. Btw, I tried the tip of the master (3 days ago) without the patch, and it resulted in lot's of warning during computation. Is it supposed to be like that?

Comment (Michoumichmich) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: @ravil-mobile do you mean compilation ? What are they ?

Comment (Michoumichmich) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: BTW, if you need a single command setup script for the whole oneAPI/DNN/MKL, I've written one here https://github.com/Michoumichmich/oneAPI-setup-script

Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Hi, @Michoumichmich. I tried `2021-09` release and it seems to that everything is working now.

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 16 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py]: Thanks for the idea. done

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 15 above)
Comment (ravil-mobile) [reply var/spack/repos/builtin/packages/dpcpp/package.py:25]: `variant('arm', ...` removed

**Comment by ravil-mobile** [comment]
Hi all, I waited a bit until I found a stable version of `dpcpp`. `2021-09` seems to work fine. I also tested with cuda@11 and it worked for me. 

I made a rebase with the `develop` branch and made some changes based on @alalazo suggestions.

File: var/spack/repos/builtin/packages/dpcpp/package.py (hunk 11 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/dpcpp/package.py:34]: This is still needed

**Review by adamjstewart — APPROVED** [review]
Looks good from my end, will let @alalazo merge once it's good on his end

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
We don't have `ninja` marked as a generator at the package level, but if need be it can be a one line fixup in a following PR.
